src/ifs/wl_seat/touch_owner.rs

```diff
@@ -0,0 +1,139 @@
+use {
+    crate::{
+        fixed::Fixed,
+        ifs::wl_seat::WlSeatGlobal,
+        tree::{FindTreeUsecase, FoundNode, Node},
+        utils::clonecell::CloneCell,
+    },
+    ahash::AHashSet,
+    std::{cell::RefCell, rc::Rc},
+};
+
+pub struct TouchOwnerHolder {
+    default: Rc<DefaultTouchOwner>,
+    owner: CloneCell<Rc<dyn TouchOwner>>,
+}
+
+impl Default for TouchOwnerHolder {
+    fn default() -> Self {
+        Self {
+            default: Rc::new(DefaultTouchOwner),
+            owner: CloneCell::new(Rc::new(DefaultTouchOwner)),
+        }
+    }
+}
+
+impl TouchOwnerHolder {
+    pub fn down(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32, x: Fixed, y: Fixed) {
+        self.owner.get().down(seat, time_usec, id, x, y)
+    }
+
+    pub fn up(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32) {
+        self.owner.get().up(seat, time_usec, id)
+    }
+
+    pub fn motion(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32, x: Fixed, y: Fixed) {
+        self.owner.get().motion(seat, time_usec, id, x, y)
+    }
+
+    pub fn frame(&self, seat: &Rc<WlSeatGlobal>) {
+        self.owner.get().frame(seat)
+    }
+
+    pub fn cancel(&self, seat: &Rc<WlSeatGlobal>) {
+        self.owner.get().cancel(seat)
+    }
+
+    pub fn clear(&self) {
+        self.owner.set(self.default.clone());
+    }
+}
+
+struct DefaultTouchOwner;
+
+struct GrabTouchOwner {
+    node: Rc<dyn Node>,
+    down_ids: RefCell<AHashSet<i32>>,
+}
+
+trait TouchOwner {
+    fn down(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32, x: Fixed, y: Fixed);
+    fn up(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32);
+    fn motion(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32, x: Fixed, y: Fixed);
+    fn frame(&self, seat: &Rc<WlSeatGlobal>);
+    fn cancel(&self, seat: &Rc<WlSeatGlobal>);
+}
+
+impl TouchOwner for DefaultTouchOwner {
+    fn down(&self, seat: &Rc<WlSeatGlobal>, time_usec: u64, id: i32, x: Fixed, y: Fixed) {
+        let mut found_tree = seat.touch_found_tree.borrow_mut();
+        let x_int = x.round_down();
+        let y_int = y.round_down();
+        found_tree.push(FoundNode {
+            node: seat.state.root.clone(),
+            x: x_int,
+            y: y_int,
+        });
+        seat.state
+            .root
+            .node_find_tree_at(x_int, y_int, &mut found_tree, FindTreeUsecase::None);
+        if let Some(node) = found_tree.last() {
```

Comment:
Touch focus has to be added to `NodeSeatState`.

Reply:
I'm not sure what to do here.  Add `touch_foci` and create new `touch_focus/touch_unfocus` functions? Or add it to `focus/unfocus`?

Reply:
`touch_foci` is what I had in mind.